src/larecs/world.mojo

```diff
@@ -756,6 +812,177 @@ struct World[*component_types: ComponentType](
+            has_start_indices=True,
+        ],
+    ) raises:
+        """
+        Adds components to multiple entities at once that are specified by a [..query.Query].
+        The provided query must ensure that matching entities do not already have one or more of the
+        components to add.
+
+        Example:
+
+        ```mojo {doctest="add_query_comps" global=true}
+        from larecs import World
+
+        @fieldwise_init
+        struct Position(Copyable, Movable):
+            var x: Float64
+            var y: Float64
+
+        @fieldwise_init
+        struct Velocity(Copyable, Movable):
+            var x: Float64
+            var y: Float64
+
+        world = World[Position, Velocity]()
+        _ = world.add_entities(Position(0, 0), 100)
+
+        for entity in world.add[Velocity](
+            world.query[Position]().without_mask[Velocity](),
+            Velocity(0.5, -0.5),
+        ):
+            velocity = entity.get[Velocity]()
+            position = entity.get[Position]()
+            entity.set[Position](Position(position.x + velocity.x, position.y + velocity.y))
+            entity.set[Velocity](Velocity(velocity.x - 0.05, velocity.y - 0.05))
+        ```
+
+        Parameters:
+            Ts: The types of the components to add.
+
+        Args:
+            query: The query specifying which entities to modify. The query parameter `has_without_mask` must always
+                be `True` to explicitly exclude entities that already have some of the components to add.
+            add_components: The components to add.
+
+        Raises:
+            Error: when called on a locked world. Do not use during [.World.query] iteration.
+            Error: when called with a query that could match entities that already have at least one of the components
+                to add.
+        """
+
+        # Note:
+        #    This operation can never map multiple archetypes onto one, due to the requirement that components to add
+        #    must be excluded in the query. Therefore, we can apply the transformation to each matching archetype
+        #    individually without checking for edge cases where multiple archetypes get merged into one.
+        #    This also enables potential parallelization optimizations.
+
+        self._assert_unlocked()
+
+        alias component_ids = Self.component_manager.get_id_arr[*Ts]()
+
+        # If query could match archetypes that already have at least one of the components, raise an error
+        if not query.without_mask[].contains(BitMask(component_ids)):
+            raise Error(
+                "Query could match entities that already have at least one of"
+                " the components to add. Use `Query.without[Component, ...]()`"
+                " to exclude those components."
+            )
+
+        arch_start_idcs = List[UInt, True](len(self._archetypes))
+        changed_archetype_idcs = List[Int, True](len(self._archetypes))
```

Comment:
It might be overly pessimistic to reserve memory for the total number of archetypes in the world. Maybe set a (generous) upper limit and risk growing the list later?

Reply:
Yes, I was thinking that the total number of archetypes will never be that big. Even `255 * 8 byte ~ 2kb` which is nothing, keeping in mind that current L1 caches have hundreds of kilobyte (e.g. see https://www.amd.com/de/support/downloads/drivers.html/processors/ryzen/ryzen-5000-series/amd-ryzen-5-5600.html).
What would your proposed initial capacity be?